buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java

```diff
@@ -196,12 +187,49 @@ private void configureStartClustersHook(Project project) {
                 @Override
                 public void beforeActions(Task task) {
                     // we only start the cluster before the actions, so we'll not start it if the task is up-to-date
-                    usedClusters.getOrDefault(task, Collections.emptyList()).stream()
+                    List<ElasticsearchCluster> neededButNotRunning = usedClusters.getOrDefault(
+                        task,
+                        Collections.emptyList()
+                    )
+                        .stream()
                         .filter(cluster -> runningClusters.contains(cluster) == false)
-                        .forEach(elasticsearchCluster -> {
-                            elasticsearchCluster.start();
-                            runningClusters.add(elasticsearchCluster);
+                        .collect(Collectors.toList());
+
+                    if (neededButNotRunning.isEmpty() == false) {
+                        int totalNodesForTask = neededButNotRunning.stream()
+                            .map(cluster -> cluster.getNumberOfNodes())
+                            .reduce(Integer::sum).get();
+                        if (totalNodesForTask > RateLimitInternalExtension.maxPermits()) {
+                            throw new TestClustersException("Clusters " + task + " are too large, requires " + totalNodesForTask +
+                                " nodes, but this system only supports " + RateLimitInternalExtension.maxPermits()
+                            );
+                        }
+                        // It seems that Gradle runs `beforeActions` and `afterExecute` in the same single worker so
+                        // blocking here would mean that all of these hooks are blocked and the permits are never
+                        // released.
+                        // To prevent deadlocking, we have to use to make use if a thread that blocks when no permits are
+                        // available.
+                        // The task will eventually block when waiting for the cluster. We don't count that towards the
+                        // timeout.
+                        RateLimitInternalExtension.executorService(project).submit( () -> {
```

Comment:
Much of this logic should really live in `RateLimitInternalExtension`. If not, all that class does is serve as a cumbersome wrapper around a semaphore. Our rate limiter should provide some kind of `lease(int, Runnable)` method and a corresponding `release()`.